DataMigration/src/main/java/org/sagebionetworks/dian/datamigration/tools/schedulev2/ScheduleV2Migration.java

```diff
@@ -0,0 +1,335 @@
+import org.sagebionetworks.dian.datamigration.tools.adherence.SageV2Availability;
+import org.sagebionetworks.dian.datamigration.tools.adherence.ScheduledSessionStart;
+import org.sagebionetworks.dian.datamigration.tools.adherence.earnings.EarningDetails;
+import org.sagebionetworks.dian.datamigration.tools.rescheduler.TestSchedule;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+import static java.util.Comparator.*;
+
+public class ScheduleV2Migration {
+
+    public static ObjectMapper objectMapper = new ObjectMapper();
+    public static Gson gson = new Gson();
+    public static SageScheduleController controller = new SageScheduleController();
+
+    public static void main(String[] args) throws IOException {
+        BridgeJavaSdkUtil.initialize(args[0], args[1], args[2]);
+        runV2Migration();
+    }
+
+    public static void runV2Migration() throws IOException {
+
+        List<Study> studyList = BridgeJavaSdkUtil.getAllStudies();
+        for (Study study : studyList) {
+            String studyId = study.getIdentifier();
+            HashSet<String> arcIdList = BridgeJavaSdkUtil.getArcIdsInStudy(studyId);
+
+            // wbfxxk is the ScheduleTemplate study, it is a test study for figuring out the V2 schedule
+            // Once we are ready to deploy this for all studies, use a curated list of all Study IDs
+            for (String arcId : arcIdList) {
+                StudyParticipant p = BridgeJavaSdkUtil.getParticipantByExternalId(arcId);
+
+                if (p.getStudyIds() == null || p.getStudyIds().isEmpty()) {
+                    System.out.println(arcId + " has withdrawn");
+                    continue; // user has withdrawn, no need to migrate
+                }
+
+                String uId = p.getId();
+                String sId = p.getStudyIds().get(0);
+
+                // Check for a user that has already migrated to V2 and signed into the app
+                SageUserClientData clientData = SageUserClientData.Companion.fromStudyParticipant(gson, p);
+                if (SageUserClientData.Companion.hasMigrated(clientData)) {
+                    System.out.println(arcId + " has already migrated to V2, leave their info alone");
+                    continue;
+                }
+
+                SageV1Schedule schedule = createV2Schedule(uId, sId);
+                if (schedule == null) {
+                    System.out.println(arcId + " has no schedule");
+                    continue;
+                }
+
+                SageV2Availability availability = createV2Availability(
+                        uId, getAvailabilityJsonFromBridge(uId));
+                if (availability == null) {
+                    System.out.println(arcId + " has no availability");
```

Comment:
Is there a situation where the availability is null, but you would still want to create adherence records?

Reply:
When a user first logs in, they can't continue in the app without having both a schedule JSON and an availability JSON.  So, if those two are not saved as a user report, they would never have been able to get into the app to take a test session.